karapace/coordinator/schema_coordinator.py

```diff
@@ -0,0 +1,935 @@
+            raise err
+        return resp
+
+    def check_errors(self) -> None:
+        """Check if coordinator is well and no authorization or unrecoverable
+        errors occurred
+        """
+        if self._coordination_task.done():
+            self._coordination_task.result()
+        if self._error_consumed_fut is not None:
+            self._error_consumed_fut.set_result(None)
+            self._error_consumed_fut = None
+        if self._pending_exception is not None:
+            exc = self._pending_exception
+            self._pending_exception = None
+            raise exc
+
+    async def _push_error_to_user(self, exc: BaseException) -> Coroutine[Any, Any, tuple[set[Any], set[Any]]]:
+        """Most critical errors are not something we can continue execution
+        without user action. Well right now we just drop the Consumer, but
+        java client would certainly be ok if we just poll another time, maybe
+        it will need to rejoin, but not fail with GroupAuthorizationFailedError
+        till the end of days...
+        XXX: Research if we can't have the same error several times. For
+             example if user gets GroupAuthorizationFailedError and adds
+             permission for the group, would Consumer work right away or would
+             still raise exception a few times?
+        """
+        exc = copy.copy(exc)
+        self._pending_exception = exc
+        self._error_consumed_fut = create_future()
+        return asyncio.wait(
+            [self._error_consumed_fut, self._closing],
+            return_when=asyncio.FIRST_COMPLETED,
+        )
+
+    async def close(self) -> None:
+        """Close the coordinator, leave the current group
+        and reset local generation/memberId."""
+        if self._closing.done():
+            return
+
+        self._closing.set_result(None)
+        # We must let the coordination task properly finish all pending work
+        if not self._coordination_task.done():
+            await self._coordination_task
+        await self._stop_heartbeat_task()
+        await self._maybe_leave_group()
+
+    def maybe_leave_group(self) -> asyncio.Task:
+        task = create_task(self._maybe_leave_group())
+        return task
+
+    async def _maybe_leave_group(self) -> None:
+        if self.generation > 0 and self.group_instance_id is None:
+            # this is a minimal effort attempt to leave the group. we do not
+            # attempt any resending if the request fails or times out.
+            # Note: do not send this leave request if we are running in static
+            # partition assignment mode (when group_instance_id has been set).
+            version = 0 if self._client.api_version < (0, 11, 0) else 1
+            request = LeaveGroupRequest[version](self.group_id, self.member_id)
+            try:
+                await self.send_req(request)
+            except Errors.KafkaError as err:
+                LOG.error("LeaveGroup request failed: %s", err)
```

Comment:
or maybe not, after a while if the consumer doesn't reply for the consumer group kafka will assign the group to another consumer, need to check the protocol to understand if the heartbeet its explicit for consumergroup-topic or if its implicit, because if its implicit we should retry this until we get a successfull response